### PR TITLE
Add ListMLE Loss

### PR DESCRIPTION
**Pull request by LakshmiKalaKadali**
Added ListMLELoss code to listwise ranking. This code does not consider Lambda weights.

### CODE REVIEW

**Comment by abheesht17** [comment]
Not sure why tests did not run, commenting so that tests run

File: keras_rs/src/losses/list_mle_loss.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+        super().__init__(**kwargs)
+
+        if temperature <= 0.0:
+            raise ValueError(
+                f"`temperature` should be a positive float. Received: "
+                f"`temperature` = {temperature}."
+            )
+
+        self.temperature = temperature
+        self._epsilon = 1e-10
```
[the remark this answers is not in the document]
Comment (LakshmiKalaKadali) [reply keras_rs/src/losses/list_mle_loss.py:66]: I am thinking — importing from keras.config.epsilon() would work, but in this case, defining epsilon locally gives us the flexibility to choose a value other than the default 1e-7.

File: keras_rs/src/losses/list_mle_loss.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        
+        num_valid_items = ops.sum(ops.cast(valid_mask, dtype=labels.dtype), 
+                                  axis=1, keepdims=True)
+        
+        batch_has_valid_items = ops.greater(num_valid_items, 0.0)
+        
+        labels_for_sorting = ops.where(valid_mask, labels, ops.full_like(labels, -1e9))
+        logits_masked = ops.where(valid_mask, logits, ops.full_like(logits, -1e9))
+        
+        sorted_indices = ops.argsort(-labels_for_sorting, axis=-1)
```
[the remark this answers is not in the document]
Comment (LakshmiKalaKadali) [reply keras_rs/src/losses/list_mle_loss.py:101]: Yes,  exactly. I have sorted manually, we can reuse the function`sort_by_scores`.

**Review by abheesht17 — COMMENTED** [review]
This LGTM! I'm going to drop an approval on this, but could you please format the code first?

**Comment by abheesht17** [comment]
Not sure why we don't have a way to run tests here, need to figure that out before merging

**Review by abheesht17 — COMMENTED** [review]
Noticed a few nits. Instead of using +, -, *, /, let's use Keras ops everywhere

**Comment by LakshmiKalaKadali** [comment]
Hi @abheesht17, 

The torch test fails are due to precision mismatch in between actual and expected values.  That precision mismatch might be due to the mask AFTER exp().  Applied the mask to set invalid positions to (-1e9) before calling ops.exp(), and removed the masking that happens after exp(). I hope that this resolves the torch test failure. Could you please run the tests. Thank You

File: keras_rs/src/losses/list_mle_loss.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        # added stable offset before calling sort_by_scores
+        list_size = ops.shape(labels_for_sorting)[1]
+        indices = ops.arange(list_size)
+
+        indices = ops.expand_dims(indices, axis=0)
+        indices = ops.broadcast_to(indices, ops.shape(labels_for_sorting))
+
+        stable_offset = ops.cast(indices, labels_for_sorting.dtype) * 1e-6
+
+        labels_for_sorting = ops.subtract(labels_for_sorting, stable_offset)
```
Comment (hertschuh) [inline keras_rs/src/losses/list_mle_loss.py]: Oh... do the issue is that Torch's `topk` is not stable with `sorted=True`, unlike JAX and TF.

There is not easy fix for `topk` that would be efficient.

However, do you mind:
- moving this in the `sort_by_scores` in `ranking_metrics_utils.py`
- doing it only if `keras.backend.backend() == "torch"` and `shuffle_ties` is False

Thanks!

**Review by hertschuh — COMMENTED** [review]
Thanks for figuring out the issue!

File: keras_rs/src/losses/list_mle_loss.py (hunk 3 above)
Comment (LakshmiKalaKadali) [reply keras_rs/src/losses/list_mle_loss.py]: Hi @hertschuh , 

That makes perfect sense. I've moved the fix to `sort_by_scores` function to handle the PyTorch top-k stability issue. I ensured that the stable sort logic is only applied when:`keras.backend.backend() == "torch" and shuffle_ties is False`
I've also added the required comment with references to the specified PyTorch issues. Thank you
